eclipse-collections/src/main/java/org/eclipse/collections/impl/set/immutable/ImmutableSetFactoryImpl.java

```diff
@@ -155,7 +157,28 @@
 
         if (Iterate.isEmpty(items))
         {
-            return this.with();
+            return this.empty();
+        }
+
+        if (items instanceof Set)
+        {
+            if (((Set) items).size() > 4)
+            {
+                return (ImmutableSet<T>) ImmutableUnifiedSet.newSet(items);
```

Comment:
I don't think this is something we should do. Generally, we want to call toArray() before constructing an immutable collection because toArray() is atomic.